### PR TITLE
refactor: introduce stream plan ref & base

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/generic.rs [hunk 1]
```diff
@@ -845,26 +845,25 @@ pub struct TopN<PlanRef> {
     pub group_key: Vec<usize>,
 }
 
-pub trait GenericBase {
-    fn schema(&self) -> &Schema;
-    fn logical_pk(&self) -> &[usize];
+pub trait GenericPlanNode {
+    fn schema(&self) -> Schema;
+    fn logical_pk(&self) -> Vec<usize>;
```
Comment (ice1000) [inline src/frontend/src/optimizer/plan_node/generic.rs:850]: Are you sure about the copy?

File: src/frontend/src/optimizer/plan_node/generic_derive.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::catalog::{Field, Schema};
+
+use super::generic::*;
+use crate::expr::{Expr, ExprDisplay};
+use crate::session::OptimizerContextRef;
+
+impl<PlanRef: GenericPlanRef> GenericPlanNode for Project<PlanRef> {
```
Comment (ice1000) [inline src/frontend/src/optimizer/plan_node/generic_derive.rs:21]: This is a cool idea!!

**Review by ice1000 — COMMENTED** [review]
Very dio

File: src/frontend/src/optimizer/plan_node/generic.rs (hunk 1 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/generic.rs:850]: yes! you can see the project's implementation, before a plannode(the inner struct) generating plan base and converting to plan ref, we can not get a reference of these property easily

**Review by ice1000 — APPROVED** [review]

**Pull request merged**